File: android/app/src/main/java/com/openburnbar/data/policy/MobileInboxSelectionPolicy.kt
package com.openburnbar.data.policy

data class MobileInboxSelectionState(
    val selectedID: String? = null,
    val pendingFocusID: String? = null,
    val filter: String = "active",
    val searchQuery: String = "",
    val focusRequestToken: Int = 0,
)

/** Cold/warm inbox focus. Source: iOS `AIInboxStore.focus` / `reconcileSelection`. */
object MobileInboxSelectionPolicy {
    fun focus(state: MobileInboxSelectionState, itemID: String?, recordIDs: List<String>): MobileInboxSelectionState {
        val pending = if (itemID != null && itemID !in recordIDs) itemID else null
        return state.copy(
            filter = "active",
            searchQuery = "",
            focusRequestToken = state.focusRequestToken + 1,
            selectedID = itemID ?: state.selectedID,
            pendingFocusID = if (itemID != null) pending else state.pendingFocusID,
        )
    }

    fun reconcile(state: MobileInboxSelectionState, visibleIDs: List<String>, recordIDs: List<String>): MobileInboxSelectionState {
        val selectedID = state.selectedID ?: return state
        if (state.pendingFocusID != null && state.pendingFocusID !in recordIDs) {
            return state
        }
        val released = if (state.pendingFocusID != null && state.pendingFocusID in recordIDs) {
            state.copy(pendingFocusID = null)
        } else {
            state
        }
        return if (selectedID !in visibleIDs) {
            released.copy(selectedID = visibleIDs.firstOrNull())
        } else {
            released
        }
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/MobileInboxSelectionPolicy.swift
import Foundation

public struct MobileInboxSelectionState: Sendable, Equatable {
    public var selectedID: String?
    public var pendingFocusID: String?
    public var filter: String
    public var searchQuery: String
    public var focusRequestToken: Int

    public init(
        selectedID: String? = nil,
        pendingFocusID: String? = nil,
        filter: String = "active",
        searchQuery: String = "",
        focusRequestToken: Int = 0
    ) {
        self.selectedID = selectedID
        self.pendingFocusID = pendingFocusID
        self.filter = filter
        self.searchQuery = searchQuery
        self.focusRequestToken = focusRequestToken
    }
}

/// Cold/warm inbox focus. Source: iOS `AIInboxStore.focus` / `reconcileSelection`.
public enum MobileInboxSelectionPolicy {
    public static func focus(
        state: MobileInboxSelectionState,
        itemID: String?,
        recordIDs: [String]
    ) -> MobileInboxSelectionState {
        var next = state
        next.filter = "active"
        next.searchQuery = ""
        next.focusRequestToken += 1
        if let itemID {
            next.selectedID = itemID
            next.pendingFocusID = recordIDs.contains(itemID) ? nil : itemID
        }
        return next
    }

    public static func reconcile(
        state: MobileInboxSelectionState,
        visibleIDs: [String],
        recordIDs: [String]
    ) -> MobileInboxSelectionState {
        var next = state
        guard let selectedID = next.selectedID else { return next }
        if let pending = next.pendingFocusID {
            guard recordIDs.contains(pending) else { return next }
            next.pendingFocusID = nil
        }
        if visibleIDs.contains(selectedID) == false {
            next.selectedID = visibleIDs.first
        }
        return next
    }
}
